agreement/message.go

```diff
@@ -17,13 +17,18 @@
 package agreement
 
 import (
+	"github.com/algorand/go-algorand/data/basics"
 	"github.com/algorand/go-algorand/protocol"
 )
 
 // A message represents an internal message which is passed between components
 // of the agreement service.
+//msgp:ignore MessageHandle
 type message struct {
-	MessageHandle
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// explicitly unexport this field since both msgp serialization and reflection randomized testing will fail since MessageHandle is interface{} type
+	messageHandle MessageHandle `codec:"-"`
```

Comment:
Yes. In practice it is messageMetadata which is a struct with no exported fields so it wouldn't have serialized it or would have serialized it with empty struct.

Reply:
This is not quite right.

```golang
func TestMessageSer(t *testing.T) {
	msg := message{
		MessageHandle: &messageMetadata{raw: network.IncomingMessage{Tag: protocol.Tag("mytag"), Data: []byte("some data")}},
		Tag:           protocol.ProposalPayloadTag,
	}

	result := protocol.EncodeReflect(&msg)
	fmt.Printf("%s", hex.Dump(result))
}
```

gives this:
```
=== RUN   TestMessageSer
00000000  89 a6 42 75 6e 64 6c 65  80 af 43 6f 6d 70 6f 75  |..Bundle..Compou|
00000010  6e 64 4d 65 73 73 61 67  65 82 a8 50 72 6f 70 6f  |ndMessage..Propo|
00000020  73 61 6c 80 a4 56 6f 74  65 80 ad 4d 65 73 73 61  |sal..Vote..Messa|
00000030  67 65 48 61 6e 64 6c 65  80 a8 50 72 6f 70 6f 73  |geHandle..Propos|
00000040  61 6c 80 a3 54 61 67 a2  50 50 b5 55 6e 61 75 74  |al..Tag.PP.Unaut|
00000050  68 65 6e 74 69 63 61 74  65 64 42 75 6e 64 6c 65  |henticatedBundle|
00000060  80 b7 55 6e 61 75 74 68  65 6e 74 69 63 61 74 65  |..Unauthenticate|
00000070  64 50 72 6f 70 6f 73 61  6c 80 b3 55 6e 61 75 74  |dProposal..Unaut|
00000080  68 65 6e 74 69 63 61 74  65 64 56 6f 74 65 80 a4  |henticatedVote..|
00000090  56 6f 74 65 80                                    |Vote.|
```
Note `MessageHandle` there.

New code gives much smaller message, but also fails to decode it with `protocol.Decode`
```
00000000  81 a3 54 61 67 a2 50 50                           |..Tag.PP|
    message_test.go:117: 
        	Error Trace:	message_test.go:117
        	Error:      	Received unexpected error:
        	            	msgpack decode error [pos 56]: no matching struct field found when decoding stream map with key MessageHandle
        	Test:       	TestMessageSer
--- FAIL: TestMessageSer (0.00s)
```

Reply:
The full test FYI
```
func TestMessageSer(t *testing.T) {
	// run on master

	// msg := message{
	// 	MessageHandle: &messageMetadata{raw: network.IncomingMessage{Tag: protocol.Tag("mytag"), Data: []byte("some data")}},
	// 	Tag:           protocol.ProposalPayloadTag,
	// }

	// result := protocol.EncodeReflect(&msg)
	// fmt.Printf("%s", hex.Dump(result))

	// err := os.WriteFile("/tmp/message", result, 0644)
	// require.NoError(t, err)

	// run on feature
	msg := message{
		messageHandle: &messageMetadata{raw: network.IncomingMessage{Tag: protocol.Tag("mytag"), Data: []byte("some data")}},
		Tag:           protocol.ProposalPayloadTag,
	}

	result := protocol.Encode(&msg)
	fmt.Printf("%s", hex.Dump(result))

	result, err := os.ReadFile("/tmp/message")
	require.NoError(t, err)

	var m2 message
	err = protocol.Decode(result, &m2)
	require.NoError(t, err)
	fmt.Printf("%v\n", m2)
}
```

Reply:
I'll push this as backward-compatibility test with b64 encoded message from master

Reply:
> Can we go back to the old MessageHandle name? I don't think codec_tester.go should dictate to us how to design our types — it works for us, not the other way around

What do you mean by go back to the old name? Go back to anonymous field in the struct? That doesn't work because we couldn't effectively define marshallers on MessageHandle. The bottom line is that MessageHandle won't in practice have meaningful data serialized. Reflect will just serialize the name with nil struct encoding which trips up msgp. Once I add the test and fallback to DecodeReflect any issues should only happen in the first round following the release and should be handled without panics and then going forward everything would use msgp

Reply:
I merged the test change and see what you mean. Both do fail with meaningful error messages. The only place that I see where messages are in serializable path is `player.Pending.Pending` which is of type `map[uint64]*messageEvent` so we do need to handle it and renaming it to unexport it won't work since EncodeReflect includes empty structs.

Reply:
what about usage like `cryptoVoteRequest` ? it embeds it and MessageHandle was available for serialization in the original code.

Reply:
FWIW I'm able to decode ping pong from master (200 entries: 45MB db file) because Pending.Pending is always empty so no messageEvents actually get encoded. They don't encode back the same with the current code but the only difference is omitempty and if I drop it it works fine.